### PR TITLE
build-tools: Add release command and support for commands that use a state machine to govern logic

### PR DESCRIPTION
**Pull request by tylerbutler**
Follow up to #11705.

- Adds an abstract base command class for commands that use state machines. This base command contains the scaffolding for commands that have internal state machines, including some logging for each state transition and handling the Init and Failed states.
- Adds various interfaces and types for working with the state machine.
- Adds the Fluid Unified Release Process finite state machine, which is used by the `release` command.

This PR also implements the `release` command, which is used to ensure that a release branch is in good condition, then walks the user through the release.

## The release process and state machine

The release process itself is described by a state machine. This is a visual representation:

image

The basic flow of the command is:

- The command starts and oclif (our CLI infrastructure) parses flags and arguments, validates them, etc.
- The command's _run_ method is called, which calls an internal _stateLoop_ function.
- That function loops through the machine states, passing each state to the command's state handler function.
- The state handler handles the states it knows about, and if it doesn't handle the state, it passes it to its parent's state handler. Thus, the parent classes are relevant to the overall implementation of the state handling. In previous iterations there was a deeper class hierarchy, but now there's just one base abstract state machine command and the release command that inherits from it.

## Reviewer guidance

I would particular like feedback on:

- How the code is organized - both at the file level and at the module level.
- It is a goal that given the name of a state and its code, it should be easy to understand what it's doing and the conditions under which it succeeds or fails. Did I succeed? The big picture is better understood using the state machine diagram.

Anything big will be punted to a follow-up PR so I can get this baseline in.

### CODE REVIEW

File: build-tools/packages/build-cli/package.json [hunk 1]
```diff
@@ -69,6 +70,7 @@
   "dependencies": {
     "@fluid-tools/version-tools": "^0.4.3000",
     "@fluidframework/build-tools": "^0.4.3000",
+    "@fluidframework/test-version-utils": ">=2.0.0-internal.1.0.0 <2.0.0-internal.1.1.0",
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/package.json:73]:
```suggestion
```

File: build-tools/packages/build-cli/package.json [hunk 2]
```diff
@@ -82,6 +83,7 @@
     "jssm": "^5.79.18",
     "npm-check-updates": "^16.0.0",
     "semver": "^7.3.7",
+    "semver-diff": "^4.0.0",
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/package.json:86]:
```suggestion
```

File: build-tools/packages/build-cli/src/base.ts [hunk 3]
```diff
@@ -53,7 +55,7 @@ export abstract class BaseCommand<T extends typeof BaseCommand.flags> extends Co
         return this.processedFlags as Partial<OutputFlags<typeof BaseCommand.flags>>;
     }
 
-    private _context: Context | undefined;
+    protected _context: Context | undefined;
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/base.ts:58]:
```suggestion
    private _context: Context | undefined;
```

**Review by sonalivdeshpande — APPROVED** [review]

File: build-tools/packages/build-cli/src/commands/release.ts [hunk 4]
```diff
@@ -0,0 +1,834 @@
...
+    async init() {
+        await super.init();
+
+        const context = await this.getContext();
+        await this.initMachineHooks();
+        const flags = this.processedFlags;
+
+        this.releaseGroup = flags.releaseGroup ?? flags.package!;
+        this.releaseVersion = context.getVersion(this.releaseGroup);
+        this.bumpType = flags.bumpType as VersionBumpType;
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/commands/release.ts:111]: Should we do validation on these to ensure they are well formed?

File: build-tools/packages/build-cli/src/commands/release.ts [hunk 5]
```diff
@@ -0,0 +1,834 @@
...
+    }
+
+    // eslint-disable-next-line complexity
+    async handleState(state: string): Promise<boolean> {
+        const context = await this.getContext();
+        let localHandled = true;
+
+        // First handle any states that we know about. If not handled here, we pass it up to the parent handler.
+        switch (state) {
+            case "AskReleaseDetails": {
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/commands/release.ts:130]: Might be nice to hoist the case-inner logic out into separate methods/functions. Would make it a bit easier to read through the switch flow without having to navigate all the case-specific implementation details. Will also give nice homes for documentation.

File: build-tools/packages/build-cli/src/commands/release.ts [hunk 6]
```diff
@@ -0,0 +1,834 @@
...
+                this.log(
+                    `Version after bump: ${bumpVersionScheme(currentVersion, this.bumpType!)}`,
+                );
+
+                this.machine.action("success");
+                break;
+            }
+
+            case "DoChecks": {
+                this.machine.action("success");
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/commands/release.ts:172]: Is this a TODO?

File: build-tools/packages/build-cli/src/commands/release.ts [hunk 7]
```diff
@@ -0,0 +1,834 @@
...
+
+            case "CheckPolicy": {
+                if (this.shouldCheckPolicy) {
+                    if (context.originalBranchName !== "main") {
+                        this.warn(
+                            "WARNING: Policy check fixes are not expected outside of main branch!  Make sure you know what you are doing.",
+                        );
+                    }
+
+                    // TODO: Call new check policy command
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/commands/release.ts:206]: I'm assuming this is still not actionable?

File: build-tools/packages/build-cli/src/commands/release.ts [hunk 8]
```diff
@@ -0,0 +1,834 @@
...
+                } else {
+                    this.warn(`Skipping installation.`);
+                }
+
+                this.machine.action("success");
+                break;
+            }
+
+            case "CheckMainNextIntegrated": {
+                // TODO: Implement this
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/commands/release.ts:390]: Is this still relevant?

File: build-tools/packages/build-cli/src/commands/release.ts [hunk 9]
```diff
@@ -0,0 +1,834 @@
...
+                break;
+            }
+
+            case "PromptToCommitBump": {
+                const prompt: InstructionalPrompt = {
+                    title: "NEED TO COMMIT LOCAL CHANGES",
+                    sections: [
+                        {
+                            title: "FIRST",
+                            message: `Commit the local changes and create a PR targeting the ${context.originalBranchName} branch.\n\nAfter the PR is merged, then the release of ${this.releaseGroup} is complete!`,
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/commands/release.ts:718]: Nit: looks like this string is duplicated below. Can we share a getter property or something?

File: build-tools/packages/build-cli/src/lib/branches.ts [hunk 10]
```diff
@@ -85,7 +85,7 @@ export function generateBumpVersionBranchName(
  */
 export function generateBumpDepsBranchName(
     bumpedDep: ReleaseGroup,
-    bumpType: VersionBumpTypeExtended,
+    bumpType: VersionBumpTypeExtended | "releasedDeps",
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/lib/branches.ts:88]: Nit: Might be better to introduce a type alias with this definition so we can document the semantic distinction between this and standard `VersionBumpTypeExtended`.

File: build-tools/packages/build-cli/src/machines/index.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+export { UnifiedReleaseMachine } from "./unifiedReleaseMachine";
+
+export { StateMachine, StateHandler } from "./types";
+
+/**
+ * An array of all known machines. Intended for testing.
+ *
+ * @internal
+ * */
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/machines/index.ts:16]:
```suggestion
 */
```

**Review by Josmithr — APPROVED** [review]

File: build-tools/packages/build-cli/src/lib/branches.ts (hunk 10 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/lib/branches.ts:88]: Agreed. Leaving type cleanup for a future PR.

File: build-tools/packages/build-cli/src/commands/release.ts (hunk 5 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/release.ts:130]: Yeah, I have started exploring that structure in another PR. Essentially using the switch statement as a "router" to the right method. This also should enable better testing of each state handler function, at least any that can be made pure functions.

File: build-tools/packages/build-cli/src/commands/release.ts (hunk 8 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/release.ts:390]: Yeah, only used in major releases, so coming in a future PR.

File: build-tools/packages/build-cli/src/commands/release.ts (hunk 7 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/release.ts:206]: Change is made, coming in future PR.

File: build-tools/packages/build-cli/src/commands/release.ts (hunk 6 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/release.ts:172]: Addressed in a pending PR.

**Pull request merged**